### PR TITLE
KAFKA-4387: Fix KafkaConsumer not responding correctly to interrupts,…

### PR DESCRIPTION
**Pull request by srdo**
... throw InterruptException from blocking methods when interrupted

See https://issues.apache.org/jira/browse/KAFKA-4387

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 1]
```diff
@@ -673,6 +675,61 @@ public void testWakeupWithFetchDataAvailable() {
...
+
+                // respond to the outstanding fetch so that we have data available on the next poll
+                client.respondFrom(fetchResponse(tp0, 0, 5), node);
+                client.poll(0, time.milliseconds());
+                
+                //Get the messages
+                consumer.poll(0);
+                
+                // Interrupt the thread and call commitSync
+                Thread.currentThread().interrupt();
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: If we're just interrupting from within the same thread that the consumer is running in, why do we need a background thread?

File: clients/src/test/java/org/apache/kafka/clients/MockClient.java [hunk 2]
```diff
@@ -65,8 +66,8 @@ public FutureResponse(Struct responseBody, boolean disconnected, RequestMatcher
     private Node node = null;
     private final Set<String> ready = new HashSet<>();
     private final Map<Node, Long> blackedOut = new HashMap<>();
-    private final Queue<ClientRequest> requests = new ArrayDeque<>();
-    private final Queue<ClientResponse> responses = new ArrayDeque<>();
+    private final Queue<ClientRequest> requests = new ConcurrentLinkedDeque<>();
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MockClient.java]: Why is this needed? I don't think we make any assumptions about the thread safety of instances of `KafkaClient`, which is why `ConsumerNetworkClient` provides its own synchronization.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java [hunk 3]
```diff
@@ -243,6 +247,9 @@ public void poll(long timeout, long now, PollCondition pollCondition) {
             // trigger wakeups after checking for disconnects so that the callbacks will be ready
             // to be fired on the next call to poll()
             maybeTriggerWakeup();
+            
+            //Throw InterruptException if this thread is interrupted
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: nit: can we follow the comment convention used in the rest of this method (i.e. lower case first word and space after the '/')? We're not super consistent globally, but it would be nice to at least be consistent locally.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 1 above)
Comment (srdo) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: If the consumer runs in the main thread, the test will hang if the test fails (if commitSync loops indefinitely). I could put a timeout on the test instead, and do away with the consumer thread?

File: clients/src/test/java/org/apache/kafka/clients/MockClient.java (hunk 2 above)
Comment (srdo) [reply clients/src/test/java/org/apache/kafka/clients/MockClient.java]: It was needed for the added test because the main thread adds a response the consumer thread then reads. I'll remove the consumer thread and revert this instead.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java (hunk 3 above)
Comment (srdo) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java]: Sure

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: I see. Alternatively, you could verify the interrupt in the call to `poll()` instead of `commitSync()`, since it already has a timeout.

Comment (srdo) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Yes, that's a better idea. Calling commitSync risks leaking the test thread if the test fails.

**Comment by hachikuji** [comment]
One other small thing. In `AbstractCoordinator`, we have a background thread which is used for sending heartbeats. It looks like we have a handler for `InterruptedException` and `RuntimeException`. I'm wondering if we should now also handle `InterruptException` and call `Thread.interrupted()` so that the behavior is consistent, though it may not matter too much since we end up closing the thread anyway. Might also be nice to be able to have the same log message in both cases. What do you think?

By the way, we'll squash automatically when we merge, so don't worry about that.

**Comment by srdo** [comment]
Good catch, we should also handle InterruptException. Calling Thread.interrupted probably doesn't matter, but it doesn't hurt to do it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java [hunk 4]
```diff
@@ -404,6 +411,12 @@ private void maybeTriggerWakeup() {
             throw new WakeupException();
         }
     }
+    
+    private void maybeThrowInterruptException() {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: One thing I'm thinking about. This method can trigger `InterruptException` in any call to `poll()`, even if the API is non-blocking (e.g. `commitAsync`). It can also be raised from `close()`, which depends on `commitSync` internally, but disables wakeups since there is often a race condition on shutdown between the call to `close()` and the call to `wakeup()`. This may be OK. We can tell users that `wakeup()` should be the preferred option for interrupting the consumer, but that `interrupt()` can be used for a "hard" shutdown. In that case, it's actually kind of nice that you can use it to interrupt a blocking `close()`. Alternatively, we could follow "wakeup-like" semantics and only check the interrupt status when wakeups are enabled. In that case, `interrupt()` in `close()` would be ignored, but we wouldn't have to worry about raising the interrupt from a non-blocking call.

One other issue. In `ConsumerCoordinator`, we have a couple places where we catch `WakeupException`, intending it to be rethrown to the user. Seems like we also need to catch `InterruptException` to avoid having the `InterruptException` swallowed by the generic handler.

Comment (srdo) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: It seems fine to me for InterruptException to be a possibility on async callbacks. When the polling thread is interrupted it can't make any progress anyway, due to the way the NIO Selector treats interrupts (returns immediately, doesn't clear interrupted state), so quitting early makes sense I think.

Ignoring interrupts while closing would reintroduce the bug this is trying to solve, since close loops forever if the polling thread is interrupted (close calls this https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L494 which never terminates if Selector just keeps returning immediately due to interrupts).

I'm not sure I understand why there is usually a race between calls to `close()` and `wakeup()`? Took a look at the KafkaConsumer Javadoc, and I can't really tell why the example at https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L420 is implemented the way it is. It seems like the `closed` AtomicBoolean is only necessary so we can log WakeupException if it happens unexpectedly. Changing the while loop to be `while (true)` seems like it would avoid the potential race.

```
public void run() {
         try {
             consumer.subscribe(Arrays.asList("topic"));
             while (true) {
                 ConsumerRecords records = consumer.poll(10000);
                 // Handle new records
             }
         } catch (WakeupException | InterruptException e) {
             // Ignore exception if closing
             if (!closed.get()) throw e;
         } finally {
             consumer.close();
         }
     }

     // Shutdown hook which can be called from a separate thread
     public void shutdown() {
         closed.set(true); // or maybe just remove this field entirely?
         consumer.wakeup(); // or interrupt, either is fine
     }
```

I'll add catches to the places where WakeupException is caught.

Comment (srdo) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: Why is catching WakeupException here https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L225 and here https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L333 necessary? I don't see anything in the try block that can throw WakeupException?

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: "Usually" was probably a poor choice of words. Users sometimes use `wakeup()` for other purposes than shutdown (e.g. to change subscriptions), in which case, an additional flag is checked to determine whether to keep going in the loop or to shutdown. The main point is that the consumer may already be in `close()` when the user calls `wakeup()`, so we opted to ignore any `wakeup()` received in `close()`. That said, this kind of usage is a little uncommon, so it wouldn't hurt to change the example in the javadoc to a pattern relying only on `wakeup()` for shutdown.

I'm leaning toward letting the interrupt semantics differ from wakeup (as implemented in the current patch) as long as we're clear in the docs that `wakeup()` is the preferred (i.e. the safest) option. Just one point of clarification: the only time we disable wakeups in a blocking operation is in `close()`, and this only really matters if auto-commit is enabled. For a direct call to the `commitSync()` API, using `wakeup()` should work as intended. So if we wanted to mimic "wakeup semantics" for interrupts, that would be the only limitation.

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: On your question about why we catch `WakeupException` in those snippets, it's because of the rebalance callback. Users will often invoke `commitSync()` from with the revocation handler, for example, or they might call one of the `seek` APIs in the assignment handler.

Comment (srdo) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: About catching `WakeupException`: Thanks, that makes sense.

About `wakeup`/`close` races: Okay, that makes sense, though I think it should usually still be possible to put the "shut down consumer thread" flag check into the `WakeupException` catch block, and have the polling loop be of the `while (true)` form (there are probably some cases I'm not thinking of). The KafkaConsumer javadoc seems pretty clear that `wakeup` is the preferred method for interrupting a consumer, but let me know if you'd like a note added somewhere. 

Regarding `close`, you're right that it's only a problem if auto-commit is enabled, but that's still not ideal, since you can then safely interrupt a consumer thread except if it's in auto-commit mode, and happens to be closing and there are offsets to commit, in which case it'll hang.

Comment (srdo) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: To clarify a bit on why interrupts and wakeup semantics are a little different: If a thread calls wakeup while the consumer thread is in `close()`, the wakeup can be ignored and the commit can be allowed to finish. If a thread calls interrupt on the consumer thread while it is in `close()`, the interrupted thread becomes unable to progress in the commit, since Selector is nonfunctional if the calling thread is interrupted. 

So we really only have two choices in that case. Either we stop trying to commit (current solution), or we figure out a way to clear the interrupt flag, finish the commit, and set the interrupted state again before returning from `close()`. I prefer the current solution, since interrupts are usually meant to stop threads as soon as possible.

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: Thanks, I think it's reasonable to make the behavior of the two a little different. It basically gives you a hard kill option, which we don't have at the moment. By the way, it looks like we catch exceptions raised from `KafkaConsumer.close()` and rethrow them as a `KafkaException`. I guess we should probably throw the `InterruptException` instead?

Comment (srdo) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:415]: Yes, that makes sense. Thanks.

**Comment by hachikuji** [comment]
LGTM. This is technically an API change, but maybe a KIP is overkill. What do you think, @ijuma and @ewencp?

**Comment by ijuma** [comment]
There is a change in behaviour for people who call `Thread.interrupt`, but that seems OK since the previous behaviour was not useful (i.e. it fixes a bug). Aside from that, is there any other behavioural change? We now declare InterruptException in the Javadoc, but since it's a subclass of KafkaException, it will still be caught by an existing generic handler.

In any case, it would be a good idea to add a note to `upgrade.html`.

**Comment by hachikuji** [comment]
Thanks @ijuma for taking a look. I agree that a short note in `upgrade.html` would be helpful.

**Comment by srdo** [comment]
Thanks for reviewing @hachikuji @ijuma. I've added a note to upgrade.html, hopefully it is what you had in mind.

File: docs/upgrade.html [hunk 5]
```diff
@@ -14,6 +14,16 @@
...
 -->
+<h4><a id="upgrade_10_2_0" href="#upgrade_10_2_0">Upgrading from 0.8.x, 0.9.x, 0.10.0.x or 0.10.1.0 to 0.10.2.0</a></h4>
+Users upgrading from versions prior to 0.10.1.0 should follow the upgrade guide <a href="#upgrade_10_1">here</a>. Users upgrading from 0.10.1.0
+can upgrade the brokers one at a time: shut down the broker, update the code, and restart it.
+<br/>
+0.10.2.0 has <a href="#upgrade_10_2_0_breaking">Potential breaking changes</a> (Please review before upgrading).
+
+<h5><a id="upgrade_10_2_0_breaking" href="#upgrade_10_2_0_breaking">Potential breaking changes in 0.10.2.0</a></h5>
+<ul>
+    <li>Several methods on the new Java consumer may now throw <code>InterruptException</code> if the calling thread is interrupted.</li>
```
Comment (ijuma) [inline docs/upgrade.html]: Thanks, this is a good start. Some suggestions:

1. InterruptException is a subclass of KafkaException so an existing exception handler for that will still work. I think it's worth mentioning this.
2. Is there a recommended way that people should use this? For example, should people catch the exception along with WakeupException?
3. What is the purpose of this change? Is it so people can do a hard interruption?

File: docs/upgrade.html [hunk 6]
```diff
@@ -22,7 +22,11 @@
 
 <h5><a id="upgrade_10_2_0_breaking" href="#upgrade_10_2_0_breaking">Potential breaking changes in 0.10.2.0</a></h5>
 <ul>
-    <li>Several methods on the new Java consumer may now throw <code>InterruptException</code> if the calling thread is interrupted.</li>
+    <li>Several methods on the new Java consumer may now throw <code>InterruptException</code> if the calling thread is interrupted.
```
Comment (hachikuji) [inline docs/upgrade.html:25]: I think our upgrade docs tend to be a bit more concise. As long as we have the relevant info in the javadocs, maybe here we can just say that the consumer's handling of interrupts has changed and that interested users should refer to javadocs.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 7]
```diff
@@ -452,6 +453,13 @@
  *     consumer.wakeup();
  * </pre>
  *
+ * Note that while methods that can throw {@link org.apache.kafka.common.errors.WakeupException} in response to {@link #wakeup()}
+ * can also throw {@link org.apache.kafka.common.errors.InterruptException} in response to the calling thread being interrupted,
+ * it is discouraged to use thread interrupts to stop/signal a consumer thread. Preferring {@link #wakeup()} will in some cases
+ * allow for a cleaner shutdown than interrupts. {@link #commitSync()} will for example be allowed to finish committing offsets
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Actually this isn't true; wakeup can also break from `commitSync`. The only case currently where the behavior is different is in `close` (if autocommit is enabled). Maybe we can just point out that a `wakeup` received during shutdown will be ignored?

Comment (srdo) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: You are right, I misremembered. Updated this, and added a missing throws to `KafkaConsumer.close`

**Comment by hachikuji** [comment]
LGTM. Thanks for the patience with the slow review cycle. I might make a couple tiny doc tweaks when I merge.
